packages/jaeger-ui/src/model/ddg/DdgEVManager.tsx

```diff
@@ -0,0 +1,152 @@
+        vertex = new DdgVertex({ key });
+        this.vertices.set(key, vertex);
+      }
+
+      this.pathElemToVertex.set(pathElem, vertex);
+      vertex.pathElems.add(pathElem);
+
+      const connectedPathElem = pathElem.focalSideNeighbor;
+      if (connectedPathElem) {
+        const connectedVertex = this.pathElemToVertex.get(connectedPathElem);
+        if (!connectedVertex) {
+          throw new Error(`Non-focal pathElem lacks connectedVertex. PathElem: ${pathElem}`);
+        }
+
+        if (!vertex[pathElem.focalSideEdgesKey].has(connectedVertex)) {
+          const newEdge = new DdgEdge(
+            pathElem.focalSideEdgesKey === 'ingressEdges'
+              ? {
+                  from: connectedVertex,
+                  to: vertex,
+                }
+              : {
+                  from: vertex,
+                  to: connectedVertex,
+                }
+          );
+          vertex[pathElem.focalSideEdgesKey].set(connectedVertex, newEdge);
+          connectedVertex[pathElem.farSideEdgesKey].set(vertex, newEdge);
+          this.edges.add(newEdge);
+        }
+      }
+    });
+  };
+
+  private removePathElems = (removeIndices: number[]) => {
+    removeIndices.forEach(removeIdx => {
+      const pathElem = this.visibilityIdxToPathElem.get(removeIdx);
+      if (!pathElem) {
+        throw new Error(`Given visibilityIdx: "${removeIdx}" that does not exist`);
+      }
+      const key = this.getVertexKey(pathElem);
+      const vertex = this.vertices.get(key);
+      if (!vertex) {
+        throw new Error(`Attempting to remove PathElem without vertex: ${pathElem}`);
+      }
+
+      this.pathElemToVertex.delete(pathElem);
+      vertex.pathElems.delete(pathElem);
+      if (vertex.pathElems.size === 0) {
+        this.vertices.delete(key);
+        vertex.egressEdges.forEach((egressEdge, connectedVertex) => {
+          connectedVertex.ingressEdges.delete(vertex);
+          this.edges.delete(egressEdge);
+        });
+
+        vertex.ingressEdges.forEach((ingressEdge, connectedVertex) => {
+          connectedVertex.egressEdges.delete(vertex);
+          this.edges.delete(ingressEdge);
+        });
+      }
+    });
+  };
+
+  // This function assumes the density is set to PPE with distinct operations
+  // It is a class property so that it can be aware of density in late-alpha
+  //
+  // It might make sense to live on PathElem so that pathElems can be compared when checking how many
+  // inbound/outbound edges are visible for a vertex, but maybe not as vertices could be densitiy-aware and
+  // provide that to this fn. could also be property on pathElem that gets set by addPathElems
+  // tl;dr may move in late-alpha
+  private getVertexKey = (pathElem: PathElem): string => {
+    const { distance, memberOf } = pathElem;
+    const { focalIdx, members } = memberOf;
+    const startIdx = Math.min(focalIdx, focalIdx + distance);
+
+    return members
+      .slice(startIdx, startIdx + Math.abs(distance) + 1)
```

Comment:
I think this could  just be `members.slice(Math.min(pathElem.memberOf.focalIdx, pathElem.memberIdx), Math.max(pathElem.memberOf.focalIdx, pathElem.memberIdx) + 1).map(/* ... */).join('|');`
a bit clearer than `startIdx` and `Math.abs(distance) + 1` IMO